lib/controllers/login.js

```diff
@@ -21,121 +19,132 @@ var loginWithOAuthProvider = require('../helpers/login-with-oauth-provider');
  * @param {Object} req - The http request.
  * @param {Object} res - The http response.
  */
-module.exports = function (req, res) {
-  var accepts = req.accepts(['html', 'json']);
+module.exports = function (req, res, next) {
   var application = req.app.get('stormpathApplication');
-  var authenticator = stormpath.OAuthPasswordGrantRequestAuthenticator(application);
   var config = req.app.get('stormpathConfig');
   var logger = req.app.get('stormpathLogger');
   var nextUri = url.parse(req.query.next || '').path;
-  var formActionUri = (config.web.login.uri + (nextUri ? ('?next=' + nextUri) : ''));
-  var view = req.app.get('stormpathConfig').web.login.view;
-  var oauthStateToken = null;
-  var providerData = req.body && req.body.providerData;
-
-  function renderForm(form, options) {
-    if (options === undefined) {
-      options = {};
+
+  res.locals.status = req.query.status;
+
+  function authenticate(options, callback) {
+    var authenticator = stormpath.OAuthPasswordGrantRequestAuthenticator(application);
+
+    options = options || {};
+    options.username = options.username || options.login || '';
+    options.password = options.password || '';
+
+    if (options.username === '' || options.password === '') {
+      return callback(new Error('Invalid username or password.'));
     }
 
-    var hasSocialProviders = _.some(config.socialProviders, function (socialProvider) {
-      return socialProvider.enabled;
-    });
+    authenticator.authenticate(options, function (err, authenticationResult) {
+      if (err) {
+        return callback(err);
+      }
 
-    extend(options, {
-      form: form,
-      formActionUri: formActionUri,
-      oauthStateToken: oauthStateToken,
-      hasSocialProviders: hasSocialProviders
-    });
+      authenticationResult.getAccount(function (err, account) {
+        if (err) {
+          logger.info('After successfully authenticating user, we were unable to retrieve the account details from Stormpath.');
+          return callback(err);
+        }
 
-    helpers.render(req, res, view, options);
-  }
+        helpers.loginResponder(authenticationResult, account, req, res);
 
-  if (req.user && config.web.login.enabled) {
-    var nextUrl = nextUri || config.web.login.nextUri;
-    return res.redirect(302, nextUrl);
+        callback();
+      });
+    });
   }
 
-  res.locals.status = req.query.status;
+  helpers.handleResponse(req, res, next, {
+    'application/json': function () {
+      var self = this;
```

Comment:
I don't see how this adds bloat, as it actually reduces your code by 1 line. And `.bind()` is a valid solution to this problem, as `self` is more like a hack.